models/target.go

```diff
@@ -0,0 +1,116 @@
+package models
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/miekg/dns"
+	"github.com/pkg/errors"
+)
+
+/* .Target is kind of a mess.
+For simple rtypes it is the record's value. (i.e. for an A record
+	it is the IP address).
+For complex rtypes (like an MX record has a preference and a value)
+	it might be a space-delimited string with all the parameters, or it
+	might just be the hostname.
+
+This was a bad design decision that I regret. Eventually we will eliminate this
+field and replace it with setters/getters.  The setters/getters are below
+so that it is easy to do things the right way in preparation.
+*/
+
+// GetTargetField returns the target. There may be other fields (for example
+// an MX record also has a .MxPreference field.
+func (rc *RecordConfig) GetTargetField() string {
+	return rc.Target
+}
+
+// GetTargetSingle returns the target for types that have a single value target
+// and panics for all others.
+func (rc *RecordConfig) GetTargetSingle() string {
+	if rc.Type == "MX" || rc.Type == "SRV" || rc.Type == "CAA" || rc.Type == "TLSA" || rc.Type == "TXT" {
+		panic("TargetSingle called on a type with a multi-parameter rtype.")
+	}
+	return rc.Target
+}
+
+// GetTargetIP returns the net.IP stored in Target.
+func (rc *RecordConfig) GetTargetIP() net.IP {
+	if rc.Type != "A" && rc.Type != "AAAA" {
+		panic(errors.Errorf("GetTargetIP called on an inappropriate rtype (%s)", rc.Type))
+	}
+	return net.ParseIP(rc.Target)
+}
+
+// GetTargetCombined returns a string with the various fields combined.
+// For example, an MX record might output `10 mx10.example.tld`.
+func (rc *RecordConfig) GetTargetCombined() string {
+	// If this is a pseudo record, just return the target.
+	if _, ok := dns.StringToType[rc.Type]; !ok {
+		return rc.Target
+	}
+
+	// We cheat by converting to a dns.RR and use the String() function.
+	// This combines all the data for us, and even does proper quoting.
+	// Sadly String() always includes a header, which we must strip out.
+	// TODO(tlim): Request the dns project add a function that returns
+	// the string without the header.
+	rr := rc.ToRR()
```

Comment:
Ug. If we are cheating by converting to RR, maybe we should bite the bullet and use RR as our internal representation. Or some type that embeds or encapsulates an RR for metadata and stuff. And maybe some additional implementations for custom stuff.

Getting kinda frustrated at the complexity we've built to avoid it.

Reply:
Agreed